sys/fido2/ctap/ctap.c

```diff
@@ -0,0 +1,1508 @@
+    uint8_t pin_token_enc[CTAP_PIN_TOKEN_SIZE] = { 0 };
+    int ret;
+    size_t sz;
+
+    if (!fido2_ctap_pin_is_set()) {
+        return CTAP2_ERR_PIN_NOT_SET;
+    }
+
+    if (!req->key_agreement_present || !req->pin_hash_enc_present) {
+        return CTAP2_ERR_MISSING_PARAMETER;
+    }
+
+    ret = fido2_ctap_crypto_ecdh(shared_secret, sizeof(shared_secret),
+                                 &req->key_agreement.pubkey);
+
+    if (ret != CTAP2_OK) {
+        return ret;
+    }
+
+    /* last moment where transaction can be cancelled */
+    if (should_cancel()) {
+        return CTAP2_ERR_KEEPALIVE_CANCEL;
+    }
+
+    /* sha256 of shared secret ((abG).x) to obtain shared key */
+    sha256(shared_secret, sizeof(shared_secret), shared_key);
+
+    sz = sizeof(pin_hash_dec);
+    ret = fido2_ctap_crypto_aes_dec(pin_hash_dec, &sz, req->pin_hash_enc,
+                                    sizeof(req->pin_hash_enc), shared_key,
+                                    sizeof(shared_key));
+
+    if (ret != CTAP2_OK) {
+        DEBUG("set pin: error while decrypting pin hash \n");
+        return ret;
+    }
+
+    if (memcmp(pin_hash_dec, g_state.pin_hash, sizeof(g_state.pin_hash)) != 0) {
+        DEBUG("fido2_ctap: get_pin_token - invalid pin \n");
+        fido2_ctap_crypto_reset_key_agreement();
+
+        ret = decrement_pin_attempts();
+
+        if (ret != CTAP2_OK) {
+            return ret;
+        }
+
+        return CTAP2_ERR_PIN_INVALID;
+    }
+
+    reset_pin_attempts();
+
+    sz = sizeof(pin_token_enc);
+    ret = fido2_ctap_crypto_aes_enc(pin_token_enc, &sz, g_pin_token,
+                                    sizeof(g_pin_token), shared_key,
+                                    sizeof(shared_key));
+
+    if (ret != CTAP2_OK) {
+        DEBUG("get pin token: error encrypting pin token \n");
+        return ret;
+    }
+
+    return fido2_ctap_cbor_encode_pin_token(encoder, pin_token_enc,
+                                            sizeof(pin_token_enc));
+}
+
+static int save_pin(uint8_t *pin, size_t len)
+{
+    uint8_t buf[SHA256_DIGEST_LENGTH] = { 0 };
+
+    sha256(pin, len, buf);
+    memmove(g_state.pin_hash, buf, sizeof(g_state.pin_hash));
+    g_state.pin_is_set = true;
+
+    save_state(&g_state);
+
+    return CTAP2_OK;
+}
+
+bool fido2_ctap_cred_params_supported(uint8_t cred_type, int32_t alg_type)
+{
+    if (cred_type == CTAP_PUB_KEY_CRED_PUB_KEY) {
+        if (alg_type == CTAP_COSE_ALG_ES256) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static inline bool pin_protocol_supported(uint8_t version)
+{
+    return version == CTAP_PIN_PROT_VER;
+}
+
+bool fido2_ctap_pin_is_set(void)
+{
+    return g_state.pin_is_set;
+}
+
+static inline bool locked(void)
+{
+    return g_state.rem_pin_att <= 0;
+}
+
+static inline bool boot_locked(void)
+{
+    return g_rem_pin_att_boot <= 0;
+}
+
+static int decrement_pin_attempts(void)
+{
+    g_state.rem_pin_att--;
+    g_rem_pin_att_boot--;
```

Comment:
Yes we do. Only using `g_state.rem_pin_att` is not possible as we need a way to know how many wrong pin attempts there have been since a reboot in order to boot lock the authenticator.

Will add guards against overflow